### DISCUSSION TITLE
ESP32 and DS3231micro

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#10810] [ESP32] [answered]
Anybody using this library for the DS3231:
https://github.com/notUnique/DS3231micro

If I use pins 5 and 4 I can talk to the DS3231, however pin5 might be a problem for the ESP32 as it is one of the strapping pins.

If I change to pins 19 and 18 I get the following error:
```py
  File "DS3231micro.py", line 66, in getSeconds
OSError: [Errno 19] ENODEV
```
Any suggestions how to fix this are appreciated.

**Comment by peterhinch** [comment, marked as answer]
See [portable DS3231 driver](https://github.com/peterhinch/micropython-samples/tree/master/DS3231).

**Reply by davefes** [reply]
I did find yours first, but as I wanted alarms to wake-up ESP32s I went with this one.  I wondered if there was some limitation on SoftI2C.  The example in the docs uses pin5 so I'll stop worrying about it's strapping pin status.

**Reply by peterhinch** [reply]
I don't believe there is any restriction on pins. Is it possible that the strapping pins work because they have pullups? If so, you could use other pins and provide your own pullups.

**Reply by davefes** [reply]
The supplied schematic showed 10K pull-ups but on closer inspection they are 4K7 AND are used for other purposes!  One goes the INT pin and the other is buried under the DS3231 chip.

Thank you for the prompt.

**Reply by peterhinch** [reply]
@davefes You might be interested in [my new DS3231 driver](https://github.com/peterhinch/micropython-samples/blob/master/DS3231/README.md). This handles alarms and is very compact. The docs explain some facts about alarm behaviour which the datasheet does not really make clear.

**Comment by davefes** [comment]
Thank you for that.  I have got version 2 of my ESPNow water tank monitoring system, with a repeater, using two DS3231 to wake-up the remote and repeater at the "same time".  I will be building another link with just a remote (one DS3231) so I'll give your library a try.

**Reply by peterhinch** [reply]
Even with DS3231 precision there could be a misalignment of 4 minutes after a year. Would this matter in your setup?

I have a system where, once an hour, a main device listens for transmissions from three remote, battery powered, devices. Each one has a one minute time slot in which they can communicate. This has run for several years, but requires accurate long term synchronisation. The main device transmits a time signal to each remote every time they connect. Each remote uses this to set their RTC. This ensures that the remotes wake up with 1 second precision - there is no long term drift even though their RTC's are less precise than a DS3231.

**Reply by davefes** [reply]
You have pretty described my system.  I get NTP time at the local and send the epoch to the repeater who then sends to the remote.  Again, an hourly slot with all the transactions over in 30 seconds.

The only important thing is that they are synced.  Why would there be drift ... do some of your devices only check-in very infrequently?

**Reply by peterhinch** [reply]
The point I was trying to make is that, however good the nodes' RTC's are, given enough time there will be some drift. This means that all nodes need periodically to re-synchronise to the master. However it sounds as if you're already on top of this :)

**Reply by davefes** [reply]
Yup, every hour.  I might cut that back to once a day and save some more battery power.
